contributors/design-proposals/lifecycle-hooks.md

```diff
@@ -0,0 +1,152 @@
+The API is open for other types of triggering conditions in the future. 
+
+### Previous Implementations
+Lifecycle hooks are already implemented in OpenShift but we want to enhance them and implement them natively in Kubernetes.
+ - [OpenShift proposal](https://github.com/openshift/origin/blob/master/docs/proposals/post-deployment-hooks.md)
+ - [OpenShift docs](https://docs.openshift.org/latest/dev_guide/deployments/deployment_strategies.html#lifecycle-hooks)
+
+### Previous Kubernetes Proposals
+ - https://github.com/kubernetes/kubernetes/pull/33545
+
+### Use Cases
+The most common use case for lifecycle hooks is to have pre, mid and/or post hook. 
+It is mostly used to run some kind of acceptance check in the middle and/or at the end to fully verify the update is working as expected and rollback if it isn't. 
+The acceptance check may be time consuming and more thorough than what readiness and liveness probes are intended for. 
+You can also notify external services from them, migrate database in the middle of an update, send messages to IRC channel or do anything else.
+
+[A short demo](https://youtu.be/GVNTm_K43iI) simulating lifecycle hooks using auto-pausing that has been presented at SIG-Apps meeting on August 21, 2017.
+
+
+## API Objects
+### New Objects
+```go
+type HookTriggerType string
+
+const (
+	ProgressAvailableHookTriggerType HookTriggerType = "ProgressAvailable"
+)
+
+type ProgressAvailableParams struct {
+    // ProgressPoint specifies the point during update when the hook should be triggered.
+    // Accepts both the number of new pods available or a percentage value representing the ratio
+    // of new.availableReplicas to the declared replicas. In case of getting a percentage
+    // it will try to reach the exact or the closest possible point right after it,
+    // trigger the job, wait for it to complete and then continue. 
+    // If such situation shall occur that two different ProgressPoints should be reached at
+    // the same time, all the hooks for an earlier ProgressPoint will be ran (and finished)
+    // before any later one.
+    ProgressPoint intstr.IntOrString
+}
+
+type HookTrigger struct {
+    // Type of hook trigger. Can be ProgressAvailable. Default is ProgressAvailable.
+    Type HookTriggerType
+    
+    // ProgressAvailable config params. Present only if type = "ProgressAvailable".
+    ProgressAvailableParams *ProgressAvailableParams
+}
+
+type LifecycleHook struct {
+    // Name of the hook
+    Name string
+    
+    // Triggering config
+    Trigger HookTrigger
+    
+    // "Abort" - Failure for this hook is fatal, deployment becomes perma-failed
+    // "Ignore" - Even if the hook fails deployment continues rolling out new version.
+    FailurePolicy string
+    
+    // "Always" - always keep this jobs 
+    // "OnFailure" - keep this Job only if it fails
+    // "Never" - delete this job immediately after it is done 
+    RetainPolicy string
+    
+    JobSpec v1.JobSpec
```

Comment:
I don't think the apps API group should depend on the batch API group in this way. What is needed more than a pod template?

Reply:
Do you really need to expose all of Job?  Wouldn't we be better to librarify the job control logic and use it from a purpose-built controller?